services/management/management.go

```diff
@@ -102,81 +112,168 @@ func NewManagement(parentCtx context.Context, config Config, provider Provider,
+		historic, err := s.provider.Get(ctx, referenceTime)
+		if err != nil {
+			return nil, errors.Wrapf(err, "management provider failed to get historic reference data for reference-time %d", referenceTime)
+		}
+		s.cachedHistoricData = historic
+		s.metrics.pageCachedStartRefTime.Update(int64(s.cachedHistoricData.StartPageReference))
+		s.metrics.pageCachedEndRefTime.Update(int64(s.cachedHistoricData.EndPageReference))
+	}
+	return s.cachedHistoricData, nil
 }
 
-func (s *service) getCommittee(refTime primitives.TimestampSeconds) *CommitteeTerm {
-	s.RLock()
-	defer s.RUnlock()
-	i := len(s.data.Committees) - 1
-	for ; i > 0 && refTime < s.data.Committees[i].AsOfReference; i-- {
+func (s *service) getData(ctx context.Context, referenceTime primitives.TimestampSeconds) (*VirtualChainManagementData, error) {
+	data, err := s.tryGetCurrentData(referenceTime)
+	if err != nil {
+		return nil, err
+	} else if data != nil {
+		return data, nil
+	}
+	data, err = s.tryGetHistoricData(ctx, referenceTime)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+/*
+ * Event-finding Functions
+ */
+func getCommittee(refTime primitives.TimestampSeconds, committees []CommitteeTerm) *CommitteeTerm {
+	i := len(committees) - 1
+	for ; i > 0 && refTime < committees[i].AsOfReference; i-- {
 	}
-	return &s.data.Committees[i]
+	return &committees[i]
 }
 
 func (s *service) GetCommittee(ctx context.Context, input *services.GetCommitteeInput) (*services.GetCommitteeOutput, error) {
+	data, err := s.getData(ctx, input.Reference)
+	if err != nil {
+		return nil, err
+	}
+
 	return &services.GetCommitteeOutput{
-		Members: s.getCommittee(input.Reference).Members,
+		Members: getCommittee(input.Reference, data.Committees).Members,
 	}, nil
 }
 
-func (s *service) getSubscriptionStatus(refTime primitives.TimestampSeconds) *SubscriptionTerm {
-	s.RLock()
-	defer s.RUnlock()
-	i := len(s.data.Subscriptions) - 1
-	for ; i > 0 && refTime < s.data.Subscriptions[i].AsOfReference; i-- {
+func getSubscriptionStatus(refTime primitives.TimestampSeconds, subscrptions []SubscriptionTerm) *SubscriptionTerm {
+	i := len(subscrptions) - 1
+	for ; i > 0 && refTime < subscrptions[i].AsOfReference; i-- {
 	}
-	return &s.data.Subscriptions[i]
+	return &subscrptions[i]
 }
 
 func (s *service) GetSubscriptionStatus(ctx context.Context, input *services.GetSubscriptionStatusInput) (*services.GetSubscriptionStatusOutput, error) {
+	data, err := s.getData(ctx, input.Reference)
+	if err != nil {
+		return nil, err
+	}
+
 	return &services.GetSubscriptionStatusOutput{
-		SubscriptionStatusIsActive: s.getSubscriptionStatus(input.Reference).IsActive,
+		SubscriptionStatusIsActive: getSubscriptionStatus(input.Reference, data.Subscriptions).IsActive,
+	}, nil
+}
+
+func getProtocolVersion(refTime primitives.TimestampSeconds, protocolVersions []ProtocolVersionTerm) *ProtocolVersionTerm {
+	i := len(protocolVersions) - 1
+	for ; i > 0 && refTime < protocolVersions[i].AsOfReference; i-- {
```

Comment:
are you sure the array covers all possible values? if the first element is not 0 you may return the first element even if refTime is before AsOfReference

Reply:
that is why there is an importance of order.

Reply:
if you have [ {asof : 50 ...}, {asof: 250 ...}, {asof 200 ..} ] the 250 one is unreachable by design. it signifies there was a version we decided to supersede  but we keep a record of it.

Reply:
i understand - but in your example if someone asks about ref 10 you will return the answer that matches 50

Reply:
?

Reply:
depends on the page. if the page is from 0 to 300 then its ok to return
in theory the page should start from earliest event.